src/platform/datapath_epoll.c

```diff
@@ -2450,12 +2450,34 @@ CxPlatSendDataSend(
             Status = QUIC_STATUS_PENDING;
         } else {
             Status = errno;
+#ifdef UDP_SEGMENT
+            QuicTraceEvent(
+                DatapathErrorStatus,
+                "[data][%p] ERROR, %u, %s.",
+                SocketContext->Binding,
+                Status,
+                "sendmsg (GSO) failed");
+#else
             QuicTraceEvent(
                 DatapathErrorStatus,
                 "[data][%p] ERROR, %u, %s.",
                 SocketContext->Binding,
                 Status,
                 "sendmmsg failed");
+#endif
+
+            if (Status == EIO) {
+                //
+                // EIO generally indicates the GSO isn't supported by the NIC,
+                // so disable segmentation on the datapath globally.
+                //
+                QuicTraceEvent(
+                    LibraryError,
+                    "[ lib] ERROR, %s.",
+                    "Disabling segmentation support globally");
+                SendData->SocketContext->Binding->Datapath->Features &=
```

Comment:
Is it safe to assume the transport will retry by nature in the common case?

It may be worth segmenting and retrying to reduce cost of transport retrying. See also the comment above about being sure UDP_SEGMENT was involved, as we wouldn't want to get stuck retrying forever.

Reply:
Yes, the transport will retry. Perhaps in the future we can consider segmenting and trying again, but it's not strictly required right now.